### ISSUE TITLE
No connection to account after building from source (SSL fail?)

### ISSUE THREAD

**Issue by mray** [dino/dino#57]
After following [those steps](https://malte-kiefer.de/post/Dino%20ein%20neuer%20XMPP%20Client/) (and using sudo all the time) Dino does not connect and permanently is caught in loop to try and fail connecting. On both servers: jabber.ccc.de and and my own [freedombox](https://github.com/freedombox) server. 

```
** (dino:1120): CRITICAL **: file /opt/dino/xmpp-vala/src/core/stanza_writer.vala: line 21: unexpected error: Unacceptable TLS certificate (g-tls-error-quark, 2)

** (dino:1120): CRITICAL **: file /opt/dino/xmpp-vala/src/core/stanza_writer.vala: line 10: uncaught error: Unacceptable TLS certificate (g-tls-error-quark, 2)

** (dino:1120): CRITICAL **: file /opt/dino/xmpp-vala/src/core/stanza_reader.vala: line 49: unexpected error: Unacceptable TLS certificate (g-tls-error-quark, 2)

** (dino:1120): CRITICAL **: file /opt/dino/xmpp-vala/src/core/stanza_reader.vala: line 49: unexpected error: Unacceptable TLS certificate (g-tls-error-quark, 2)
Stream Error: Content before root node
Gtk-Message: GtkDialog mapped without a transient parent. This is discouraged.

```

Looks like this is an SSL related thing.

**Comment by mar-v-in** [comment]
jabber.ccc.de uses CACert which is most likely not activated as a CA on your system and thus there is no way to authenticate the remote server. Your "own freedombox" likely has a similar issue.

Dino currently does not allow to ignore or handle certificate issues. If you trust CACert you can install their CA to your system or use a XMPP server that does not use broken certificates.

**Comment by wolkenschieber** [comment]
Same issue occurs, when the jabber server uses a self-signed certificate.

**Comment by kehugter** [comment]
Just build Dino and found this issue too. Right now the UI just loops trying to connect, without any message to the user. In future binary releases, it'd be ideal to have some meaningful alert for the user and break the log in attempt.

**Comment by ouafnico** [comment]
Hi
Any fix ?

**Comment by andypl** [comment]
Hey guys,

I have the same problem on my hosted domain @ conversations.im server. It use Let's Encrypt certificate.

Gajim, Conversations and https://check.messaging.one are fine with it but Dino give me an error: Stream Error: TlsError: Unacceptable TLS certificate

**Comment by nifker** [comment]
Sad that this is still an issue...

**Comment by metbosch** [comment]
Just in case someone reaches this point in the future.

I modified the login form to contain an additional `servername` field. This allows defining a different domain for the server than the one in the JID. I required this modification as my company has the server running at `jabber.company.com` but the JID must be `metbosch@company.com` to make the LDAP integration correctly work.

The code is in the master branch of my fork: https://github.com/metbosch/dino/
The error that the client was showing to me is:

> Its security certificate is issued to another domain.

**Comment by mar-v-in** [comment]
@metbosch @liacon-kter
These changes won't be merged as they are RFC non-compliant.

The problem with metbosch's company server is that it uses a TLS certificate for jabber.company.com to claim ownership of user@company.com, which is not supposed to work. The proposed change does not only change the server to connect to, it also changes the TLS verification such that it allows a certificate from that server to authenticate for any domain entered in JID field, which is wrong. 

Even if a company for some strange reason is unable to use SRV records for company.com and thus requires their clients to enter jabber.company.com as a host to connect to, the TLS certificate presented must be for the company.com domain to indicate ownership of user@company.com address.

**Comment by metbosch** [comment]
@licaon-kter @mar-v-in 

I can to the PR if it is interesting for the community. I didn't open it before because I do not control the corner cases (like leaving the field in blank) and I do not want to put to much effort on that. My idea was to leave the information here, just in case someone founds it interesting or it can help to develop the real solution.

Regarding @mar-v-in comment, I do not know if the mismatch between JID and servername is protocol compliant or not, but other jabber clients support the definition of both fields separately.

Here is an screenshot of my modified sing in window:
![Screenshot from 2019-10-21 10-52-36](https://user-images.githubusercontent.com/4162025/67190800-ed460380-f3f0-11e9-9d5e-b64a6213a134.png)

**Comment by mar-v-in** [comment]
I am not against the option to set custom server name and port (although I won't agree with the proposed UI), however setting a custom server still requires verifying the TLS certificate against the domain part of the JID.

You mentioned that you see the "Its security certificate is issued to another domain." error when connecting to your JID without entering the server name. This is a strong indication that Dino already knows the correct server to connect to and the server name and port setting should be useless in your case. Your patch helps not because it allows you to connect to your server but because it allows you to accept an illegal TLS certificate.

This can be a severe security threat to non-expert users, as they might read in the internet that entering `connect.repair` into that box will fix any connection issues. As they connect the server at `connect.repair` will receive their login credentials and is able to drive a MITM attack.

This is why a server must be able to present a TLS certificate for the domain part of the JID so that no matter what you add as a server name, it is verified that the server is actually authenticated to receive your JIDs login credentials.

**Comment by hinneLinks** [comment]
Workaround (Ubuntu 18.04): Add your self-signed Cert to the System's Certstore:


1. Get the Certificate as a .crt (Or PEM and Rename it). I [tried](https://stackoverflow.com/questions/54576731/get-certificate-of-xmpp-s2s-gateway) to download it myself, but that didn't work (I got the Servers cert, not one for my Domain, see below). I downloaded it via https://xmpp.net there you can analyze your Server, in the Result you can Download the Certificate as a .pem File. Download it and rename it to .crt.
1a. If you used Gajim before, you can check the downloaded Certs-SHA-256-Key: Find the Key Gajim used in ~/.config/gajim/config , search for "ssl_fingerprint_sha256". Compare that to the output of `openssl x509 -in /the/crt/file/you/downloaded.crt -noout -text -sha256 -fingerprint`
2. Add the Cert as described [here](https://unix.stackexchange.com/a/132163/243679):
`sudo cp my.crt /usr/local/share/ca-certificates/`
`sudo update-ca-certificates`
3. Restart Dino, you should be able to Login now, without Errors.

Why I need to do this?
I'm using a professional Email Service, they also host XMPP. They of course have valid Certs, but I opted to use my own domain (hollerith.info). My Domain has a SRV-Record, that redirects xmpp to xmpp.mailbox.org. They issue a Self-Signed Cert for "hollerith.info", since they can't get an official cert for the Domain, because the domain is under my Control. Of course theoretically, that would be solvable, but sadly, they offer XMPP just as it is, without further support.

**Comment by YarmoM** [comment]
Any real solution in sight yet? I'm using a Let's Encrypt certificate like the rest of the tinkering world. I'm reading the different issues here but fail to see why this is still an issue after three years while all other XMPP clients work with my server?

**Comment by licaon-kter** [comment]
@YarmoM what's the issue if you use Let's Encrypt then?

**Comment by YarmoM** [comment]
Even when using Let's Encrypt, I get the "Could not establish a secure connection". Dino is supposed to work with LE certs?

**Comment by licaon-kter** [comment]
@YarmoM of course, works for most of the network.

But... you have a cert for the domain or xmpp.domain or what? Detail your cert setup...and the format of your JID

**Comment by YarmoM** [comment]
It's never fun when people use the wrong github issue to discuss their own problem. And I've just become one of them ( ,_,) sorry
Your comment made me realise my real issue. My JID is name@domain.com and the XMPP is server is located at domain.com. But the website domain.com is not hosted on the same machine as the XMPP server. I have an LE certificate for the website, but since the XMPP machine does not have access to that certificate, no wonder Dino complains. So I guess I will try to get a second certificate on the XMPP machine.

**Comment by ElDifinitivo** [comment]
This is an issue for anyone (like me) running an XMPP server as a Tor hidden service.

Let's Encrypt will not work with *.onion domains, so I created a self-signed certificate which creates 0 problems with `profanity` on desktop and `conversations` on Android. I have some people on my server who would rather use a GUI program, and this issue stops them from being able to use `dino`.

**Comment by pkonow** [comment]
Same problem for me, using my own domain with the mailbox.org XMPP service. With Pidgin and Gajim it works fine. Dino does not seem to offer a way to deal with this self-signed certificate. I would expect a dialog to pop up that asks me whether I want to accept this self-signed certificate or not. Adding the certificate to the system store does not seem to be a clean solution because I don't want to _globally_ trust this certificate, only with one specific application.

The used setup instructions can be found here: https://kb.mailbox.org/display/MAILBOX/Using+Jabber+service+with+your+own+domain

**Comment by licaon-kter** [comment]
@pkonow so POSH support? https://tools.ietf.org/html/draft-ietf-xmpp-posh-06

**Comment by pkonow** [comment]
Well, POSH might help in some specific cases, but this is more of a general issue. Any user-facing application that deals with certificates should give the user an option to manually accept self-signed certs. There can be many reasons for that, for example when connecting to your own server that nobody else uses. The application should not ultimately require a valid trust chain because there is simply no need for it in certain cases.

What's wrong with just presenting the certificate to the user, and if he trusts it then saving the fingerprint in a list of manually approved certificates? AFAIK this is how all the other XMPP clients, web browsers etc. solve it. :wink:  POSH would more be an improvement on top, to avoid requiring a manual trust statement by using an alternative method to verify the authenticity of the certificate IF such a method is available.

**Comment by licaon-kter** [comment]
POSH covers your case with hosted XMPP more broadly. But I agree.

**Comment by grondinm** [comment]
I'm running into this issue using a wildcard certificate from Let's Encrypt. the cert is for *.domain.org and domain.org the XMPP server is at chat.domain.org. Does Dino really not like the fact that it's a wildcard cert? Why would this be considered invalid?

**Comment by licaon-kter** [comment]
And the accounts are `user@chat.domain.org` or `user@domain.org` ?

**Comment by grondinm** [comment]
The accounts are user@chat.domain.org

**Comment by licaon-kter** [comment]
Nothing useful in logs? https://github.com/dino/dino/wiki/Debugging

**Comment by fiaxh** [comment]
@grondinm Dino doesn't do anything certificate related itself; that's done by glib and your system decides what's accepted. Thus, something is probably wrongly set up on your end. If you tell me your domain**, I can have a look at why Dino doesn't connect to it.
** Here or in a private message to fiaxh in the project channel `chat@dino.im`.

**Comment by SergeyKisar** [comment]
@fiaxh 
Mate, can u help me as well? I am using ejabberd under ovpn, didnt have domain name, connect directly to ip. Its fine to every client except Dino. In these conditions of couse i cant use good cert.
I am willing to pay for your time and actual result. Contact me at Telegram - Cal_013, or by email - serg-kisar@yandex.ru

**Comment by PoneyClairDeLune** [comment]
Wow, this actually still is an ongoing problem. Great to hear that I'm locked out of access to `xmpp.ilita.i2p` due to this certificate check that cannot be disabled.

Please just offer an option to disable certificate checks.
![image](https://github.com/dino/dino/assets/88174309/03422083-7bc8-4d7d-89f1-41e110c6f8c5)

**Comment by licaon-kter** [comment]
@PoneyClairDeLune as the message says you can have your **system** trust it first, right?

**Comment by PoneyClairDeLune** [comment]
> @PoneyClairDeLune as the message says you can have your **system** trust it first, right?

It's a chore for every single XMPP server hosted on mixnets. Why not just offer an option for all hosts, or if the TLD provided is one of the mixnet TLDs (`onion`, `i2p`, `loki`, etc)?

**Comment by licaon-kter** [comment]
@PoneyClairDeLune there's no such issue on .onion domains, maybe see that commit and open a PR for .i2p ?

**Comment by PoneyClairDeLune** [comment]
@licaon-kter I don't know anything about Vala. Syntactically, it looks like C. Does it have an OR operator (`||`)?

**Comment by nyovaya** [comment]
How does it come that my browser says my certificate is valid and the connection is secure but Dino says the same certificate for the XMPP server is not valid (it has the same SHA256 sum)?

**Comment by Flashwalker** [comment]
so if you have your server in local network
openfire.server.local

and you just want to connect via vpn with your jid
user@server.local

you can't even create a profile in app
and also you get a cert error/prohibition

what a useless app 😆

**Comment by pasnox** [comment]
I can't imagine this issue still running since 2017, what useless choice.
It's common some server use self signed certificate or being temporary down because of renewal issues, then you stuck users out, what a bad choice..

**Comment by dhalucario** [comment]
This kinda sucks. My provider offers a XMPP server with your own Domain via SRV DNS entries, but uses self-signed certificates. Dino won't be a choice for me if there isn't at least a "trust anyway" button.

**Comment by licaon-kter** [comment]
@dhalucario what's their use case?

**Comment by mar-v-in** [comment]
@dhalucario Some servers will reject communication with your server if it can't present a valid certificate (e.g. if [`s2s_secure_auth`](https://prosody.im/doc/s2s#security) is set in Prosody's configuration). And there is never a "trust anyway" button on the server-to-server connection, so you will not be able to receive messages from all servers if you don't have a valid and trusted certificate.

You can only reliably use invalid/self-signed or untrusted certificates in non-federated setups, which are not a primary usecase of Dino. If you really want to do it anyway, you can import the self-signed certificate or custom CA into your system's trusted certificate store, which Dino will pick up.

Research has shown that "trust anyway" buttons are almost always a bad idea, because most people will just click that button - just like most people accept all cookies when browsing the internet - without evaluating security or privacy implications.

**Comment by dhalucario** [comment]
> [@dhalucario](https://github.com/dhalucario) Some servers will reject communication with your server if it can't present a valid certificate (e.g. if [`s2s_secure_auth`](https://prosody.im/doc/s2s#security) is set in Prosody's configuration). And there is never a "trust anyway" button on the server-to-server connection, so you will not be able to receive messages from all servers if you don't have a valid and trusted certificate.
> 
> You can only reliably use invalid/self-signed or untrusted certificates in non-federated setups, which are not a primary usecase of Dino. If you really want to do it anyway, you can import the self-signed certificate or custom CA into your system's trusted certificate store, which Dino will pick up.
> 
> Research has shown that "trust anyway" buttons are almost always a bad idea, because most people will just click that button - just like most people accept all cookies when browsing the internet - without evaluating security or privacy implications.

Huh, I am not sure if my setup worked as intended then:
My hoster mailbox.org allows me to link my domain to their server via SRV records. When I tried to log into the account via my custom domain it did not work. Using the mailbox.org account worked just fine though.

**Comment by michaelblyons** [comment]
> Huh, I am not sure if my setup worked as intended then: My hoster mailbox.org allows me to link my domain to their server via SRV records. When I tried to log into the account via my custom domain it did not work. Using the mailbox.org account worked just fine though.

It is working to the degree that Mailbox.org supports. See https://github.com/dino/dino/issues/57#issuecomment-573438534 for a workaround.

I think there's another way to get the certificate,[^1] but I can't remember how. I think it was buried, encoded, in an XML file somewhere.

[^1]:  The [SO link](https://stackoverflow.com/a/56137662/241211) did not work. It just gives me the main Mailbox.org certificate.

**Comment by aehimself** [comment]
While this issue does not affect me at all, I just would like to share my thoughts.

> [@dhalucario](https://github.com/dhalucario) Some servers will reject communication with your server if it can't present a valid certificate (e.g. if [`s2s_secure_auth`](https://prosody.im/doc/s2s#security) is set in Prosody's configuration).

Especially since Chat Control is at large we'll see more and more, self-hosted XMPP "islands" pop up where server-to-server connection is not used and allowed. Maybe the server won't even have Internet access.
Yes, automating a Let's Encrypt certificate is easy and works fine with eJabberd for example, some people will not go for it. Some will simply use their company's existing self-signed certificates.

> [@dhalucario](https://github.com/dhalucario) Research has shown that "trust anyway" buttons are almost always a bad idea, because most people will just click that button - just like most people accept all cookies when browsing the internet - without evaluating security or privacy implications.

Just like "removing the french language pack" from Linux, which is still the main target of Dino. While can be dangerous it's still allowed.

I would like to stress again, I don't want to change anyone's mind... just would like to point out that the request does make sense IRL scenarios and would be a huge plus for Dino.

**Comment by git2fa** [comment]
Please refer to this duplicate bug report: https://github.com/dino/dino/issues/1768

I really want to use Dino, but this discussion here **since 8 years** shows, that the developers of Dino only want to support public servers and certificates and suppress own private servers, specially without internet connection.
**Why?**
Pidgin shows that it is no matter to inform the user that there is an "untrusted" certificate and then works with it after confirmation.

**Comment by mar-v-in** [comment]
As was already explained, research has shown that users cannot cope with TLS certificate warnings in a reasonable way. Studies suggest that
- up to 70% of users ignored TLS certificate warnings without any further verification when browsing the web
- less than 50% understood that ignoring the TLS certificate warning when checking their email could result in third-parties getting access to their email
- less than 20% could accurately identify data at risk when ignoring a TLS certificate warning
- more than 50% of users believed that ignoring TLS certificate warnings on high risk websites like banks is less of an issue, because they assumed those websites deploy better security
- users of Linux (and macOS) are more likely to ignore phishing warnings in browsers, arguing they are less at risk because there is less malware for their operating system

tl;dr: Allowing users to bypass TLS certificate warnings easily (just one or two clicks) is a security bad practice. The route we take (to require the user to import their certificate in their operating system trusted certificate store) significantly reduces the risk, as doing so is typically too complex for users that don't know what they are doing.

**Comment by git2fa** [comment]
Thank you for our response and explanation [mar-v-in](https://github.com/mar-v-in).

> Studies suggest that ...

Why do you worry about (governmental) studies and not about what the users really need?

> tl;dr: .... The route we take (to require the user to import their certificate in their operating system trusted certificate store) significantly reduces the risk, as **doing so is typically too complex for users that don't know what they are doing**.

That's the essence of the problem!
It is too complex for normal users to import a certificate into the operating system.
The only alternative is to avoid using an TLS-secured connection and you want to argue that this is more save.

It's a pity, at least Dino maybe a very good application for XMPP chat's, but it can only be used with certified public servers.
I was interested in Dino, because it is one of the rare clients that supports OMEMO.

https://github.com/dino/dino/issues/1768#issuecomment-3359960106
> As is described in the other issue, to trust a certificate you need to trust it from your operating system. Dino does not have its own trust store for certificates.

For myself I didn't realize that an own trust store is needed for the application to handle self signed certificates.
So I would appreciate to reopen https://github.com/dino/dino/issues/1768 as a wish for an approvement for Dino.

But I will try to import the certificate and test if this will work. When I have success I will report it.

**Comment by mar-v-in** [comment]
> Why do you worry about (governmental) studies and not about what the users really need?

These independent user studies merely confirm what we already know, that the majority of users will just click continue and do whatever is necessary to make the thing "work", ignoring security and privacy aspects. This is not exclusive to TLS certificate warnings, it equally applies to malware and phishing warnings, cookie banners and privacy terms.

Remember that you are not the average user. And maybe you are also not the target audience of Dino and another client, e.g. Gajim, is the better option for you.

We prefer protecting the majority of our users that have a valid TLS certificate and would click through any TLS certificate warning if it was possible to bypass it rather than adding an easy option for the few that don't. Remember that valid TLS certificates are already required for server-to-server connections.

We've added exceptions to the requirement of valid TLS certificates for use-cases where they're not critical for security, notably Tor hidden services (which have an additional transport layer protection and thus TLS security is not relevant)

> The only alternative is to avoid using an TLS-secured connection and you want to argue that this is more save.

Dino doesn't support non-TLS-secured connections either.

**Comment by ll3N1GmAll** [comment]
Not having the ability to connect to your own private server because "trust anyway" buttons cause security issues is asinine. Even connecting directly to a private IP gives this error. You're telling me a private server not exposed to anything outside of the LAN requires this level of "protection"? People don't need more infantalizing nanny-state behavior in our lives. Make great software and let people use it how they choose. Don't exclude necessary capabilities that literally every other client I have ever seen allows you to do because of potential interconnectivity issues or security concerns. If you want to warn people, by all means, do so; but let people make their own choices with their own infrastructure. The use case of needing a temporary self-signed certificate while official certs are being renewed is an excellent example of why "trust anyway" buttons are necessary. Even those doing it right will need exceptions every now and then. And not allowing connectivity to private, unroutable, IP subnets is insane. You CANNOT get a CA issued TLS cert for a 192.168.x.x address. Let's Encrypt isnt issuing certs to assets it cannot connect to. 
Devs, please allow the TLS cert verification to be turned off. At least for servers being connected to direct on the LAN via an unroutable IP subnet.

**Comment by M-II-R** [comment]
Hello! Same issue. Just wanted to test Dino UX with XMPP server running in local network and couldn't because of lack of this feature (I really don't want to generate certificate, install ant delete it just for this test). Dino looks good on screenshots I've found in the Web, but I didn't see it in the action. Profanity and Psi on Unix-like systems, Conversations on Android (and, I suppose, many other clients) can handle servers without trusted TLS certificates. In my opinion, it's always better when you can do X with some app, especially when lots of competitors have this function - it will grant you advantage over some another clients without function and remove your disadvantage in comparison with clients having it. What about adding power user menu, or CLI option, or config file activating this? I think, if the user goes to documentation site (or issues list of the repo), this will take enough time for average user to think about if he/she needs it.

**Comment by hrxi** [comment]
A power user can probably patch out the certificate check by themselves if they understand the risks.

Adding a menu, CLI option, config file just compromises on end-user security of people running unmodified Dino.
